ACHNBrowserUI/ACHNBrowserUI/views/categories/CategoriesView.swift

```diff
@@ -67,30 +67,50 @@ struct CategoriesView: View {
                     .padding(.top, 8)
                     .padding(.bottom, 4)
                     .background(Color.grassBackground)
-                List {
-                    if viewModel.searchText.isEmpty {
-                        makeNatureCell()
-                        makeWardrobeCell()
-                        makeCategories()
-                    } else {
-                        ForEach(viewModel.searchResults.keys.map{ $0 }, id: \.self) { key in
-                            Section(header: self.makeSearchCategoryHeader(category: key)) {
-                                ForEach(self.viewModel.searchResults[key] ?? []) { item in
-                                    NavigationLink(destination: ItemDetailView(item: item)) {
-                                        ItemRowView(item: item)
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }.modifier(DismissingKeyboardOnSwipe())
+                if viewModel.searchText.isEmpty {
+                    categoriesList
+                } else {
+                    searchList
+                }
             }
             .navigationBarTitle(Text("Catalog"), displayMode: .inline)
             .background(Color.grassBackground)
             
             ItemsListView(viewModel: ItemsViewModel(category: .housewares))
         }
     }
+
+    private var categoriesList: some View {
+        List {
+            makeNatureCell()
+            makeWardrobeCell()
+            makeCategories()
+        }.modifier(DismissingKeyboardOnSwipe())
+    }
+
+    private var searchList: some View {
+        List {
+            ForEach(searchCategories, id: \.0, content: searchSection)
+        }.modifier(DismissingKeyboardOnSwipe())
+    }
+
+    private var searchCategories: [(Category, [Item])] {
+        viewModel.searchResults
+            .map { $0 }
+            .sorted(by: \.key.rawValue)
```

Comment:
Try to remove this line, and expect some crashes... Oh my... I spent so much time to figure it out it was that... I'm pretty sure this problem will happen again, the crash message is really not obvious, but when you know the cause, the message makes sense!

Reply:
Make total sense actually. Didn't even realize we lost the order at each query haha.